chips/sam4l/src/usbc/common_register.rs

```diff
@@ -0,0 +1,244 @@
+#![allow(dead_code)]
+
+use core::marker::PhantomData;
+
+/// A memory-mapped register
+#[derive(Copy, Clone)]
+pub struct Reg<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> Reg<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        Reg { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn read(self) -> T {
+        T::from_word(self.read_word())
+    }
+
+    #[inline]
+    pub fn write(self, val: T) {
+        unsafe { ::core::ptr::write_volatile(self.addr, T::to_word(val)); }
+    }
+
+    #[inline]
+    pub fn read_word(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.addr) }
+    }
+}
+
+impl Reg<u32> {
+    #[inline]
+    pub fn set_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w | bit);
+    }
+
+    #[inline]
+    pub fn clr_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w & (!bit));
+    }
+}
+
+/// A write-only memory-mapped register, where any zero bits written have no effect
+pub struct RegW<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> RegW<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        RegW { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn write(self, val: T) {
+        unsafe { ::core::ptr::write_volatile(self.addr, T::to_word(val)); }
+    }
+}
+
+impl RegW<u32> {
+    #[inline]
+    pub fn set_bit(self, bit_index: u32) {
+        // For this kind of register, reads always return zero
+        // and zero bits have no effect, so we simply write the
+        // single bit requested.
+        let bit = 1 << bit_index;
+        self.write(bit);
+    }
+}
+
+/// A read-only memory-mapped register
+pub struct RegR<T> {
+    addr: *const u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> RegR<T> {
+    pub const fn new(addr: *const u32) -> Self {
+        RegR { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn read(self) -> T {
+        T::from_word(self.read_word())
+    }
+
+    #[inline]
+    pub fn read_word(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.addr) }
+    }
+}
+
+/// An array of memory-mapped registers
+pub struct Regs<T> {
```

Comment:
Aren't these plural types (`Regs`, `RegsW`, `RegsR`) extraneous since you can just use a fixed-size array? e.g. `[RegR; 22]`? It compiles to the same thing (fixed-sized arrays don't have a length field since the length is deduced for bounds checking etc at compile-time).